internal/pkg/service/common/etcdop/iterator/iterator.go

```diff
@@ -0,0 +1,55 @@
+package iterator
+
+import (
+	"context"
+
+	etcd "go.etcd.io/etcd/client/v3"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/serde"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+type Definition[R any] struct {
+	config
+}
+
+type Iterator[R any] struct {
+	config
+	ctx    context.Context
+	client *etcd.Client
+	page   int
+	index  int
+	err    error
+}
+
+func New[R any](start string, serde serde.Serde, opts ...Option) Definition[R] {
+	// Apply options
+	c := config{limit: DefaultLimit, start: start, serde: serde}
+	for _, o := range opts {
+		o(&c)
+	}
+	return Definition[R]{config: c}
+}
+
+func (v Definition[T]) Do(ctx context.Context, client *etcd.Client) *Iterator[T] {
+	return &Iterator[T]{ctx: ctx, client: client, config: v.config, page: -1, index: -1}
+}
+
+func (v *Iterator[T]) Next() bool {
+	return false
+}
+
+func (v *Iterator[T]) Value() op.KeyValueT[T] {
+	if v.index == -1 {
+		panic(errors.New("unexpected Value() call: Next() must be called first"))
+	}
+	if v.err != nil {
+		panic(errors.Errorf("unexpected Value() call: %w", v.err))
+	}
+	return op.KeyValueT[T]{}
+}
+
+func (v *Iterator[T]) Err() error {
+	return v.err
+}
```

Comment:
Najprv som si urcil ako to ma fungovat.

Vybral som si tento sposob:

https://blog.kowalczyk.info/article/1Bkr/3-ways-to-iterate-in-go.html
![image](https://user-images.githubusercontent.com/19371734/205250270-8dde6caf-11a7-4dbc-9d90-b62b730bf631.png)


Moznosti je vela, mohla by byt napr. jedna `Next()` s 3 return parameters (`value, found, err`).

Reply:
Jasně, tohle nebyl problém. 😅 (btw. Viděl jsem i nějaký benchmarky, kde právě doporučovali stateful iterátory tohoto typu oproti channelům a closures který jsou v Go výrazně pomalejší.)

Reply:
Ano.

Ratam ze do `~10000` hodnot za sekundu tam nebude rozdiel, ... pri vacsich cislach su potom `channels` pomalsie.
Teda nie je to nas use-case, ... my budeme mat tolko klucov iba pri importe `records` do file storage, kde sa to bude optimalizovat samostatne.

Callback opat nie je meratelny problem pod `<10000`.

V etcdop mame `benchmarks`, ktore som spustil pre `main` a potom pre tento PR.
`go test -v -benchmem -benchtime=1000x -bench=. ./internal/pkg/service/common/...`

A nie je tam nejaky zasadny rozdiel v rychlosti alebo pocte mem alloc.
Samozrejme nacitanie `10` stranok namiesto `1` (`1000` zaznamov po `100`) ma nejaku reziu naviac.